lensless/unrolled_fista.py

```diff
@@ -0,0 +1,129 @@
+# #############################################################################
+# unrolled_fista.py
+# =================
+# Authors :
+# Yohann PERRON [yohann.perron@gmail.com]
+# #############################################################################
+
+import numpy as np
+from lensless.trainable_recon import TrainableReconstructionAlgorithm
+from lensless.gradient_descent import non_neg
+import inspect
+from scipy import fft
+
+try:
+    import torch
+
+except ImportError:
+    raise ImportError("Pytorch is require to use trainable reconstruction algorithms.")
+
+
+class UnrolledFISTA(TrainableReconstructionAlgorithm):
+    """
+    Object for applying unrolled projected gradient descent with FISTA (Fast Iterative
+    Shrinkage-Thresholding Algorithm) for acceleration.
+
+    FISTA Paper: https://www.ceremade.dauphine.fr/~carlier/FISTA
+
+    """
+
+    def __init__(self, psf, n_iter=5, dtype=None, proj=non_neg, learn_tk=True, tk=1, **kwargs):
+        """
+        COnstructor for unrolled FISTA algorithm.
+
+        Parameters
+        ----------
+        psf : :py:class:`~torch.Tensor` of shape (H, W, C)
+            The point spread function (PSF) that models forward propagation.
+        n_iter : int, optional
+            Number of iterations to unrolled, by default 5
+        dtype : float32 or float64
+            Data type to use for optimization.
+        proj : :py:class:`function`, optional
+            Projection function to apply at each iteration, by default non_neg
+        learn_tk : bool, optional
+            whether the tk parameters of FISTA should be learnt, by default True
+        tk : int, optional
+            Initial value of tk, by default 1
+        """
+
+        super(UnrolledFISTA, self).__init__(psf, n_iter=n_iter, dtype=dtype, **kwargs)
+
+        self._proj = proj
+
+        # initial guess, half intensity image
+        # for online approach could use last reconstruction
+        psf_flat = self._psf.reshape(-1, self._n_channels)
+        pixel_start = (torch.max(psf_flat, axis=0).values + torch.min(psf_flat, axis=0).values) / 2
+        self._image_init = torch.ones_like(self._psf) * pixel_start
+        self._image_est = self._image_init
+        self._xk = self._image_init
+
+        # learnable step size initialize as < 2 / lipschitz
+        Hadj_flat = self._convolver._Hadj.reshape(-1, self._n_channels)
+        H_flat = self._convolver._H.reshape(-1, self._n_channels)
+        self._alpha = torch.nn.Parameter(
```

Comment:
allow user to specify `alpha`? otherwise default to value provided here. and if specified, add a check that it is less than 2 / lipschitz

Reply:
similar to ADMM, should we do clipping of alpha so that it is below 2 / lipchitz and non-negative?

Reply:
agree about not clipping below 2/L